File: practice/746. Min Cost Climbing Stairs.go
package main

import "fmt"

func minCostClimbingStairs(cost []int) int {
	size := len(cost)
	if size == 2 {
		return min(cost[0], cost[1])
	}

	dp := make([]int, size+1)
	dp[0] = 0
	dp[1] = 0

	for i := 2; i <= size; i++ {
		dp[i] = min(dp[i-2]+cost[i-2], dp[i-1]+cost[i-1])
	}

	return dp[size]
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(minCostClimbingStairs([]int{10, 15, 20}))
}

File: 746.min-cost-climbing-stairs.ts
/*
 * @lc app=leetcode id=746 lang=typescript
 *
 * [746] Min Cost Climbing Stairs
 */

// @lc code=start
function minCostClimbingStairs(cost: number[]): number {
  const n = cost.length;
  const dp = Array(n + 1).fill(0);

  for (let i = 2; i <= n; i++) {
    dp[i] = Math.min(dp[i - 1] + cost[i - 1], dp[i - 2] + cost[i - 2]);
  }

  return dp[n];
}
// @lc code=end
